{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from rich import print\n",
    "\n",
    "data_path = \"/home/halleluyah/Documents/hivmut/data/processed/hiv_lanl_nigeria.csv\"\n",
    "\n",
    "df = pd.read_csv(data_path, index_col=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SE id(SSAM)</th>\n",
       "      <th>PAT id(SSAM)</th>\n",
       "      <th>Name</th>\n",
       "      <th>Locus Name</th>\n",
       "      <th>Isolate Name</th>\n",
       "      <th>Clone Name</th>\n",
       "      <th>Georegion</th>\n",
       "      <th>Country</th>\n",
       "      <th>Sampling City</th>\n",
       "      <th>Sampling Year</th>\n",
       "      <th>...</th>\n",
       "      <th>Project</th>\n",
       "      <th>Patient ethnicity</th>\n",
       "      <th>Progression</th>\n",
       "      <th># of patient seqs</th>\n",
       "      <th># of patient timepoints</th>\n",
       "      <th>Species</th>\n",
       "      <th>SE id(SA)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GI number</th>\n",
       "      <th>Version</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pq223</td>\n",
       "      <td>EU269016</td>\n",
       "      <td>pq223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5222</td>\n",
       "      <td>EU269016</td>\n",
       "      <td>166831812</td>\n",
       "      <td>EU269016.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pq222</td>\n",
       "      <td>EU269015</td>\n",
       "      <td>pq222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5223</td>\n",
       "      <td>EU269015</td>\n",
       "      <td>166831810</td>\n",
       "      <td>EU269015.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5224</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pq220</td>\n",
       "      <td>EU269014</td>\n",
       "      <td>pq220</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5224</td>\n",
       "      <td>EU269014</td>\n",
       "      <td>166831808</td>\n",
       "      <td>EU269014.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pq217</td>\n",
       "      <td>EU269013</td>\n",
       "      <td>pq217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5225</td>\n",
       "      <td>EU269013</td>\n",
       "      <td>166831806</td>\n",
       "      <td>EU269013.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5226</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pq214</td>\n",
       "      <td>EU269012</td>\n",
       "      <td>pq214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5226</td>\n",
       "      <td>EU269012</td>\n",
       "      <td>166831804</td>\n",
       "      <td>EU269012.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   SE id(SSAM)  PAT id(SSAM)   Name Locus Name Isolate Name Clone Name  \\\n",
       "#                                                                        \n",
       "1         5222           NaN  pq223   EU269016        pq223        NaN   \n",
       "2         5223           NaN  pq222   EU269015        pq222        NaN   \n",
       "3         5224           NaN  pq220   EU269014        pq220        NaN   \n",
       "4         5225           NaN  pq217   EU269013        pq217        NaN   \n",
       "5         5226           NaN  pq214   EU269012        pq214        NaN   \n",
       "\n",
       "  Georegion  Country Sampling City  Sampling Year  ...  Project  \\\n",
       "#                                                  ...            \n",
       "1   AFR SSA  NIGERIA           NaN            NaN  ...      NaN   \n",
       "2   AFR SSA  NIGERIA           NaN            NaN  ...      NaN   \n",
       "3   AFR SSA  NIGERIA           NaN            NaN  ...      NaN   \n",
       "4   AFR SSA  NIGERIA           NaN            NaN  ...      NaN   \n",
       "5   AFR SSA  NIGERIA           NaN            NaN  ...      NaN   \n",
       "\n",
       "  Patient ethnicity Progression # of patient seqs # of patient timepoints  \\\n",
       "#                                                                           \n",
       "1               NaN         NaN               NaN                     NaN   \n",
       "2               NaN         NaN               NaN                     NaN   \n",
       "3               NaN         NaN               NaN                     NaN   \n",
       "4               NaN         NaN               NaN                     NaN   \n",
       "5               NaN         NaN               NaN                     NaN   \n",
       "\n",
       "  Species SE id(SA) Accession  GI number     Version  \n",
       "#                                                     \n",
       "1     NaN      5222  EU269016  166831812  EU269016.1  \n",
       "2     NaN      5223  EU269015  166831810  EU269015.1  \n",
       "3     NaN      5224  EU269014  166831808  EU269014.1  \n",
       "4     NaN      5225  EU269013  166831806  EU269013.1  \n",
       "5     NaN      5226  EU269012  166831804  EU269012.1  \n",
       "\n",
       "[5 rows x 66 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800080; text-decoration-color: #800080; font-weight: bold\">Index</span><span style=\"font-weight: bold\">([</span><span style=\"color: #008000; text-decoration-color: #008000\">'SE id(SSAM)'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'PAT id(SSAM)'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Name'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Locus Name'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Isolate Name'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Clone Name'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Georegion'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Country'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Sampling City'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Sampling Year'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Sampling Year Upper'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Patient Age'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Patient Health'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Organism'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Subtype'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Phenotype'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Coreceptor'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Sample Tissue'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Culture Method'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Molecule type'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Drug Naive'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Problematic Sequence'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Viral load'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'CD4 count'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'CD8 count'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Days from Infection'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Days from Seroconversion'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Days from first Sample'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Sequencing method'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Amplification strategy'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Fiebig Stage'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Anno'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Days from treatment start'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Days from treatment end'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Vaccine status'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'RIP subtype'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Se ID'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Sequence Length'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'GB comment'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'DB comment'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Sequence'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'GB create date'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'GB update date'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Percent non-ACGT'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Map image(SE id)'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'HXB2/MAC239 start'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'HXB2/MAC239 stop'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Patient Id'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Patient Code'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Patient Sex'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Risk Factor'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Infection Country'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Infection City'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Infection Year'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Patient comment'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'HLA type'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Project'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'Patient ethnicity'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Progression'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'# of patient seqs'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'# of patient timepoints'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Species'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'SE id(SA)'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Accession'</span>,\n",
       "       <span style=\"color: #008000; text-decoration-color: #008000\">'GI number'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'Version'</span><span style=\"font-weight: bold\">]</span>,\n",
       "      <span style=\"color: #808000; text-decoration-color: #808000\">dtype</span>=<span style=\"color: #008000; text-decoration-color: #008000\">'object'</span><span style=\"font-weight: bold\">)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;35mIndex\u001b[0m\u001b[1m(\u001b[0m\u001b[1m[\u001b[0m\u001b[32m'SE id\u001b[0m\u001b[32m(\u001b[0m\u001b[32mSSAM\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m, \u001b[32m'PAT id\u001b[0m\u001b[32m(\u001b[0m\u001b[32mSSAM\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m, \u001b[32m'Name'\u001b[0m, \u001b[32m'Locus Name'\u001b[0m, \u001b[32m'Isolate Name'\u001b[0m,\n",
       "       \u001b[32m'Clone Name'\u001b[0m, \u001b[32m'Georegion'\u001b[0m, \u001b[32m'Country'\u001b[0m, \u001b[32m'Sampling City'\u001b[0m, \u001b[32m'Sampling Year'\u001b[0m,\n",
       "       \u001b[32m'Sampling Year Upper'\u001b[0m, \u001b[32m'Patient Age'\u001b[0m, \u001b[32m'Patient Health'\u001b[0m, \u001b[32m'Organism'\u001b[0m,\n",
       "       \u001b[32m'Subtype'\u001b[0m, \u001b[32m'Phenotype'\u001b[0m, \u001b[32m'Coreceptor'\u001b[0m, \u001b[32m'Sample Tissue'\u001b[0m, \u001b[32m'Culture Method'\u001b[0m,\n",
       "       \u001b[32m'Molecule type'\u001b[0m, \u001b[32m'Drug Naive'\u001b[0m, \u001b[32m'Problematic Sequence'\u001b[0m, \u001b[32m'Viral load'\u001b[0m,\n",
       "       \u001b[32m'CD4 count'\u001b[0m, \u001b[32m'CD8 count'\u001b[0m, \u001b[32m'Days from Infection'\u001b[0m,\n",
       "       \u001b[32m'Days from Seroconversion'\u001b[0m, \u001b[32m'Days from first Sample'\u001b[0m,\n",
       "       \u001b[32m'Sequencing method'\u001b[0m, \u001b[32m'Amplification strategy'\u001b[0m, \u001b[32m'Fiebig Stage'\u001b[0m, \u001b[32m'Anno'\u001b[0m,\n",
       "       \u001b[32m'Days from treatment start'\u001b[0m, \u001b[32m'Days from treatment end'\u001b[0m,\n",
       "       \u001b[32m'Vaccine status'\u001b[0m, \u001b[32m'RIP subtype'\u001b[0m, \u001b[32m'Se ID'\u001b[0m, \u001b[32m'Sequence Length'\u001b[0m,\n",
       "       \u001b[32m'GB comment'\u001b[0m, \u001b[32m'DB comment'\u001b[0m, \u001b[32m'Sequence'\u001b[0m, \u001b[32m'GB create date'\u001b[0m,\n",
       "       \u001b[32m'GB update date'\u001b[0m, \u001b[32m'Percent non-ACGT'\u001b[0m, \u001b[32m'Map image\u001b[0m\u001b[32m(\u001b[0m\u001b[32mSE id\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "       \u001b[32m'HXB2/MAC239 start'\u001b[0m, \u001b[32m'HXB2/MAC239 stop'\u001b[0m, \u001b[32m'Patient Id'\u001b[0m, \u001b[32m'Patient Code'\u001b[0m,\n",
       "       \u001b[32m'Patient Sex'\u001b[0m, \u001b[32m'Risk Factor'\u001b[0m, \u001b[32m'Infection Country'\u001b[0m, \u001b[32m'Infection City'\u001b[0m,\n",
       "       \u001b[32m'Infection Year'\u001b[0m, \u001b[32m'Patient comment'\u001b[0m, \u001b[32m'HLA type'\u001b[0m, \u001b[32m'Project'\u001b[0m,\n",
       "       \u001b[32m'Patient ethnicity'\u001b[0m, \u001b[32m'Progression'\u001b[0m, \u001b[32m'# of patient seqs'\u001b[0m,\n",
       "       \u001b[32m'# of patient timepoints'\u001b[0m, \u001b[32m'Species'\u001b[0m, \u001b[32m'SE id\u001b[0m\u001b[32m(\u001b[0m\u001b[32mSA\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m, \u001b[32m'Accession'\u001b[0m,\n",
       "       \u001b[32m'GI number'\u001b[0m, \u001b[32m'Version'\u001b[0m\u001b[1m]\u001b[0m,\n",
       "      \u001b[33mdtype\u001b[0m=\u001b[32m'object'\u001b[0m\u001b[1m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Country</th>\n",
       "      <th>Georegion</th>\n",
       "      <th>Sampling City</th>\n",
       "      <th>Sampling Year</th>\n",
       "      <th>Patient Age</th>\n",
       "      <th>Patient Health</th>\n",
       "      <th>Patient Sex</th>\n",
       "      <th>Risk Factor</th>\n",
       "      <th>...</th>\n",
       "      <th>Days from first Sample</th>\n",
       "      <th>Sequencing method</th>\n",
       "      <th>Amplification strategy</th>\n",
       "      <th>Problematic Sequence</th>\n",
       "      <th>Fiebig Stage</th>\n",
       "      <th>RIP subtype</th>\n",
       "      <th>GI number</th>\n",
       "      <th>GB comment</th>\n",
       "      <th>DB comment</th>\n",
       "      <th>Accession</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cttcagaacagaccagagccaacagccccacgaccagagccaacag...</td>\n",
       "      <td>EU269016</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>166831812</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EU269016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cttcagaacaggccagagccaacagcccagaacaggatagagccaa...</td>\n",
       "      <td>EU269015</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>166831810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EU269015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cttcaggacaggctagagccaacagccccaccagcagagagcttcg...</td>\n",
       "      <td>EU269014</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>166831808</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EU269014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ctccagaacaggctagagccaacagccccaccagcagagagcttcg...</td>\n",
       "      <td>EU269013</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>166831806</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EU269013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cttcagagcaggccagaaccaacagccccaccagcagagagcttcg...</td>\n",
       "      <td>EU269012</td>\n",
       "      <td>NIGERIA</td>\n",
       "      <td>AFR SSA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>166831804</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EU269012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Sequence  Accession  Country  \\\n",
       "#                                                                          \n",
       "1  cttcagaacagaccagagccaacagccccacgaccagagccaacag...   EU269016  NIGERIA   \n",
       "2  cttcagaacaggccagagccaacagcccagaacaggatagagccaa...   EU269015  NIGERIA   \n",
       "3  cttcaggacaggctagagccaacagccccaccagcagagagcttcg...   EU269014  NIGERIA   \n",
       "4  ctccagaacaggctagagccaacagccccaccagcagagagcttcg...   EU269013  NIGERIA   \n",
       "5  cttcagagcaggccagaaccaacagccccaccagcagagagcttcg...   EU269012  NIGERIA   \n",
       "\n",
       "  Georegion Sampling City  Sampling Year Patient Age Patient Health  \\\n",
       "#                                                                     \n",
       "1   AFR SSA           NaN            NaN         NaN            NaN   \n",
       "2   AFR SSA           NaN            NaN         NaN            NaN   \n",
       "3   AFR SSA           NaN            NaN         NaN            NaN   \n",
       "4   AFR SSA           NaN            NaN         NaN            NaN   \n",
       "5   AFR SSA           NaN            NaN         NaN            NaN   \n",
       "\n",
       "  Patient Sex Risk Factor  ...  Days from first Sample  Sequencing method  \\\n",
       "#                          ...                                              \n",
       "1         NaN         NaN  ...                     NaN                NaN   \n",
       "2         NaN         NaN  ...                     NaN                NaN   \n",
       "3         NaN         NaN  ...                     NaN                NaN   \n",
       "4         NaN         NaN  ...                     NaN                NaN   \n",
       "5         NaN         NaN  ...                     NaN                NaN   \n",
       "\n",
       "   Amplification strategy Problematic Sequence Fiebig Stage RIP subtype  \\\n",
       "#                                                                         \n",
       "1                     NaN                    0          NaN         NaN   \n",
       "2                     NaN                    0          NaN         NaN   \n",
       "3                     NaN                    0          NaN         NaN   \n",
       "4                     NaN                    0          NaN         NaN   \n",
       "5                     NaN                    0          NaN         NaN   \n",
       "\n",
       "   GI number  GB comment DB comment  Accession  \n",
       "#                                               \n",
       "1  166831812         NaN        NaN   EU269016  \n",
       "2  166831810         NaN        NaN   EU269015  \n",
       "3  166831808         NaN        NaN   EU269014  \n",
       "4  166831806         NaN        NaN   EU269013  \n",
       "5  166831804         NaN        NaN   EU269012  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List of important columns\n",
    "important_columns = [\n",
    "    'Sequence', 'Accession', 'Country', 'Georegion', 'Sampling City', 'Sampling Year',\n",
    "    'Patient Age', 'Patient Health', 'Patient Sex', 'Risk Factor', 'Infection Year',\n",
    "    'CD4 count', 'Viral load', 'Subtype', 'Phenotype', 'Coreceptor', 'Drug Naive',\n",
    "    'Days from Infection', 'Days from Seroconversion', 'Days from first Sample',\n",
    "    'Sequencing method', 'Amplification strategy', 'Problematic Sequence', 'Fiebig Stage',\n",
    "    'RIP subtype', 'GI number', 'GB comment', 'DB comment', 'Accession'\n",
    "]\n",
    "\n",
    "df_important = df[important_columns]\n",
    "df_important.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_important.to_csv('../data/processed/hiv_lanl.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions = df['Accession']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'biopython'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mbiopython\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'biopython'"
     ]
    }
   ],
   "source": [
    "import biopython\n",
    "\n",
    "# Download the reference genome for HIV-1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "bioinformatics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
